master/buildbot/configurators/janitor.py

```diff
@@ -39,43 +40,67 @@ class LogChunksJanitor(BuildStep):
     def run(self):
-        older_than_timestamp = datetime2epoch(now() - self.logHorizon)
-        deleted = yield self.master.db.logs.deleteOldLogChunks(older_than_timestamp)
-        self.descriptionDone = ["deleted", str(deleted), "logchunks"]
+        if self.logHorizon is not None:
+            older_than_timestamp = datetime2epoch(
+                now() - self.logHorizon)
+            deleted = yield self.master.db.logs.deleteOldLogChunks(
+                older_than_timestamp=older_than_timestamp)
+            self.descriptionDone = ["deleted", str(deleted), "logchunks"]
+
+        if self.horizonPerBuilder is not None:
+            deleted = yield self.master.db.logs.deleteOldLogChunks(
+                horizonPerBuilder=self.horizonPerBuilder)
+            self.descriptionDone = ["deleted", str(deleted), "logchunks"]
+
         return SUCCESS
 
 
 class BuildDataJanitor(BuildStep):
     name = 'BuildDataJanitor'
     renderables = ["build_data_horizon"]
 
-    def __init__(self, build_data_horizon):
+    def __init__(self, build_data_horizon=None, horizonPerBuilder=None):
         super().__init__()
         self.build_data_horizon = build_data_horizon
+        self.horizonPerBuilder = horizonPerBuilder
 
     @defer.inlineCallbacks
     def run(self):
-        older_than_timestamp = datetime2epoch(now() - self.build_data_horizon)
-        deleted = yield self.master.db.build_data.deleteOldBuildData(older_than_timestamp)
-        self.descriptionDone = ["deleted", str(deleted), "build data key-value pairs"]
+        if self.build_data_horizon is not None:
+            older_than_timestamp = datetime2epoch(now() - self.build_data_horizon)
+            deleted = yield self.master.db.build_data.deleteOldBuildData(
+                older_than_timestamp=older_than_timestamp)
+            self.descriptionDone = ["deleted", str(deleted), "build data key-value pairs"]
+
+        if self.horizonPerBuilder is not None:
+            deleted = yield self.master.db.build_data.deleteOldBuildData(
+                horizonPerBuilder=self.horizonPerBuilder)
+            self.descriptionDone = ["deleted", str(deleted), "build data key-value pairs"]
         return SUCCESS
 
 
 class JanitorConfigurator(ConfiguratorBase):
     """ Janitor is a configurator which create a Janitor Builder with all needed Janitor steps"""
 
-    def __init__(self, logHorizon=None, hour=0, build_data_horizon=None, **kwargs):
+    def __init__(self, logHorizon=None, hour=0, build_data_horizon=None, horizonPerBuilder=None,
+    **kwargs):
         super().__init__()
         self.logHorizon = logHorizon
         self.build_data_horizon = build_data_horizon
+        self.horizonPerBuilder = horizonPerBuilder
         self.hour = hour
         self.kwargs = kwargs
+        if ((self.logHorizon is not None or self.build_data_horizon is not None)
+        and self.horizonPerBuilder is not None):
+            bbconfig.error("JanitorConfigurator: horizonPerBuilder only " +
```

Comment:
I think it makes sense to support all of logHorizon, build_data_horizon, log_horizon_per_builder and build_data_horizon_per_builder at the same time.

Having both global and per-builder log horizons makes sense because it's then possible to have a longer global horizon and shorter per-builder horizons. It also doesn't increase the complexity of the code.